File: verification/verify_visual.py
from playwright.sync_api import sync_playwright

def run_cuj(page):
    # Mock window.prompt and avoid fetching actual external CDNs blocking render
    page.add_init_script("""
        window.prompt = () => 'mock_token';
        window.state = {
            user: { $id: 'user123', name: 'TestUser', email: 'test@example.com' },
            profile: { $id: 'prof123', username: 'TestUser', rank: 'Наблюдатель', flower_xp: 50, ether: 5 },
            profileCache: new Map(),
            aiCooldown: false
        };
        // Ensure Three.js and AvatarSystem don't crash if loading out of order
        window.addEventListener('load', () => {
             if (window.AvatarSystem) {
                 window.AvatarSystem.init();
                 // Add some mock placeholders so the animate loop has something to process
                 const container = document.getElementById('messages-container');
                 if (container) {
                     for(let i=0; i<3; i++) {
                         const ph = document.createElement('div');
                         ph.className = 'soul-avatar-placeholder';
                         ph.dataset.user = 'user' + i;
                         ph.style.width = '50px';
                         ph.style.height = '50px';
                         container.appendChild(ph);
                     }
                 }
             }
        });
    """)

    # Mock Appwrite API to avoid auth hang
    page.route('**/v1/account*', lambda route: route.fulfill(status=200, json={"$id": "user123", "name": "TestUser", "email": "test@example.com"}))
    page.route('**/v1/databases/*/collections/*/documents*', lambda route: route.fulfill(status=200, json={"documents": [{"$id": "msg1", "senderId": "TestUser", "messageContent": "Hello World", "timestamp": "2023-01-01T00:00:00.000Z"}]}))
    page.route('**/v1/realtime*', lambda route: route.abort())

    # Do NOT block unpkg/jsdelivr as Three.js is loaded from there in index.html

    page.goto("http://localhost:8080/index.html")
    page.wait_for_timeout(1000)

    # Bypass landing page to enter app
    page.evaluate("""
        document.querySelectorAll('section:not(#app-interface)').forEach(el => el.classList.add('hidden'));
        document.getElementById('app-interface').classList.remove('hidden');
        if (window.AvatarSystem && !window.AvatarSystem.isRunning) {
            window.AvatarSystem.init();
        }
    """)

    page.wait_for_timeout(2000) # wait for avatars to render

    # Take screenshot
    page.screenshot(path="verification/screenshots/verification.png")
    page.wait_for_timeout(1000)

if __name__ == "__main__":
    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        context = browser.new_context(
            record_video_dir="verification/videos"
        )
        page = context.new_page()
        try:
            run_cuj(page)
        finally:
            context.close()
            browser.close()

File: verification/verify_visual.js
const { chromium } = require('playwright');
const path = require('path');
const fs = require('fs');

(async () => {
    // 1. Setup Browser
    const browser = await chromium.launch({
        args: ['--no-sandbox', '--disable-setuid-sandbox', '--disable-web-security']
    });
    const context = await browser.newContext({
        viewport: { width: 1280, height: 720 },
        deviceScaleFactor: 1,
    });
    const page = await context.newPage();

    // 2. Mock Appwrite & Three.js to prevent errors and network calls
    await page.addInitScript(() => {
        window.Appwrite = {
            Client: class { setEndpoint() { return this; } setProject() { return this; } subscribe() {} },
            Account: class { get() { return Promise.resolve({ name: 'TestUser', email: 'test@example.com' }); } },
            Databases: class {
                listDocuments() { return Promise.resolve({ documents: [] }); }
                createDocument() { return Promise.resolve({}); }
                updateDocument() { return Promise.resolve({}); }
                deleteDocument() { return Promise.resolve({}); }
            },
            Storage: class { getFileView() { return ''; } },
            ID: { unique: () => 'unique_id' },
            Query: { orderAsc: () => {}, limit: () => {}, equal: () => {} }
        };
    });

    // 3. Load Page
    const filePath = path.resolve(__dirname, '../index.html');
    const fileUrl = 'file://' + filePath;
    console.log(`Loading: ${fileUrl}`);
    await page.goto(fileUrl, { waitUntil: 'domcontentloaded' });

    // 4. Inject Test Data (Simulate Chat with Avatars)
    await page.waitForTimeout(2000); // Allow scripts to load

    await page.evaluate(() => {
        // Force show interface
        document.getElementById('app-interface').classList.remove('hidden');
        document.getElementById('app-interface').style.display = 'block';

        const container = document.getElementById('messages-container');
        if (!container) return;

        container.innerHTML = '';

        // Add avatars
        for(let i=0; i<5; i++) {
            const row = document.createElement('div');
            row.className = 'message-row';
            row.style.height = '100px';
            row.style.width = '100%';
            row.style.display = 'flex';

            const avatar = document.createElement('div');
            avatar.className = 'soul-avatar-placeholder';
            avatar.dataset.user = `User${i}`;
            avatar.style.width = '50px';
            avatar.style.height = '50px';
            avatar.style.background = 'red'; // Visual debug
            avatar.style.border = '2px solid black';

            row.appendChild(avatar);
            container.appendChild(row);
        }

        // Initialize manually if needed
        if (window.AvatarSystem) {
             console.log("Initializing AvatarSystem manually...");
             window.AvatarSystem.init();
        }
    });

    // Wait for IntersectionObserver
    await page.waitForTimeout(2000);

    // 5. Verify Logic via Internal State
    const result = await page.evaluate(() => {
        const system = window.AvatarSystem;
        if (!system) return { error: "AvatarSystem not found" };

        const visibleCount = system.visibleAvatars ? system.visibleAvatars.size : -1;

        return {
            visibleCount,
            isRunning: system.isRunning,
            hasObserver: !!system.observer
        };
    });

    console.log('Verification State:', result);

    // 6. Screenshot for manual visual check
    await page.screenshot({ path: 'verification/avatar_verification.png', fullPage: true });
    console.log('Screenshot saved to verification/avatar_verification.png');

    await browser.close();

    if (result.visibleCount > 0 && result.hasObserver) {
        console.log('SUCCESS: Avatars are tracked via IntersectionObserver.');
        process.exit(0);
    } else {
        console.error('FAILURE: Avatar tracking failed.');
        process.exit(1);
    }
})();
